pkg/util/deployment/deployment.go

```diff
@@ -48,6 +48,39 @@ const (
 	RollbackDone              = "DeploymentRollback"
 )
 
+// GetAllReplicaSets list all RSes of a given deployment atomically
+func GetAllReplicaSets(deployment *extensions.Deployment, c clientset.Interface) (allOldRSes []*extensions.ReplicaSet, newRS *extensions.ReplicaSet, err error) {
+	namespace := deployment.Namespace
+	selector, err := unversioned.LabelSelectorAsSelector(deployment.Spec.Selector)
+	if err != nil {
+		return nil, nil, err
+	}
+	options := api.ListOptions{LabelSelector: selector}
+	rsList, err := c.Extensions().ReplicaSets(namespace).List(options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	oldRSes := map[string]extensions.ReplicaSet{}
+	newRSTemplate := GetNewReplicaSetTemplate(deployment)
+	for _, rs := range rsList.Items {
+		if err != nil {
+			return nil, nil, fmt.Errorf("invalid label selector: %v", err)
+		}
+		// Filter out replica set that has the same pod template spec as the deployment - that is the new replica set.
+		if api.Semantic.DeepEqual(rs.Spec.Template, newRSTemplate) {
+			newRS = &rs
+			continue
+		}
+		oldRSes[rs.ObjectMeta.Name] = rs
```

Comment:
There shouldn't be multiple ReplicaSets with the same name in the same namespace. The system guarantees spatial uniqueness.